File: cpp/203. Remove Linked List Elements.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;
typedef vector<int> vi;
typedef vector<vector<int>> vvi;
typedef pair<int, int> pi;


struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
// Time: O(n), Space: O(1)
public:
    ListNode* removeElements(ListNode* head, int val) {
        if(!head) return head;       

        ListNode dummy = ListNode(0);
        dummy.next = head;
        ListNode *pre = &dummy, *cur = head;

        while(cur) {
            if(cur->val == val) {
                pre->next = cur->next;
                delete cur;
                cur = pre->next;
            } else {
                pre = pre->next;
                cur = cur->next;
            }
        }
        return dummy.next;
    }
};

int main() {
    ios::sync_with_stdio(0); 
    cin.tie(0);

    
}

File: python/203-Remove Linked List Elements.py
from linklist_helper import *

class Solution:
    def removeElements(self, head, val: int) -> ListNode:
        "Time: O(n), Space: O(1)"
        dummy = ListNode()
        dummy.next = head
        prev = dummy
        cur = head
        while cur:
            if cur.val == val:
                prev.next = cur.next
            else:
                prev = cur
            cur = cur.next
        return dummy.next

head = gen_linklist([1,2,6,3,4,5,6])
val = 6
result = Solution().removeElements(head, val)
display_linklist(result)

head = gen_linklist([1,1])
val = 1
result = Solution().removeElements(head, val)
display_linklist(result)